Feat: JWT and feature flags

rdong333 opened the pull request.

Commit:
testing JWT with spring security

Commit:
WIP JWT auth porting

Commit:
porting more JWT

Commit:
trying to port jwt

Commit:
fixed syntax

Commit:
jwt port wip, refresh token failing

Commit:
jwt port wip: fixing tests

Commit:
Finished JWT porting

shawn-sher committed:
First pass at updated from jwt-feature-flag

Commit:
set feature flags in property file

Commit:
locally working JWT feature flags

Commit:
remove comment

melanahammel commented on cerberus-domain/src/main/java/com/nike/cerberus/domain/CerberusAuthToken.java:
```diff
@@ -19,16 +19,9 @@
-@AllArgsConstructor
-@NoArgsConstructor
+
 public class CerberusAuthToken implements Serializable {
```
This code is fine, but you could use Lombak's @Getter and @Setter notations to condense this class. It might not work for all the methods, but could work for some.

melanahammel replied:
https://projectlombok.org/features/Data

shawn-sher replied:
I added lombok.Getter for the gets. I wish this was not in experimental: https://www.projectlombok.org/features/experimental/Accessors

melanahammel commented on cerberus-web/src/main/java/com/nike/cerberus/jobs/JwtBlocklistRefreshJob.java:
```diff
@@ -0,0 +1,45 @@
+
+  @Scheduled(cron = "${cerberus.jobs.jwtBlocklistRefreshJob.cronExpression}")
+  public void execute() {
+    log.debug("Running JWT blocklist refresh job");
```
Should this be a log.info?

melanahammel commented on cerberus-web/src/main/java/com/nike/cerberus/jobs/JwtSecretRefreshJob.java:
```diff
@@ -0,0 +1,45 @@
+
+  @Scheduled(cron = "${cerberus.jobs.jwtSecretRefreshJob.cronExpression}")
+  public void execute() {
+    log.debug("Running JWT secret refresh job");
```
Again, should this be log.info?

melanahammel commented on cerberus-web/src/main/java/com/nike/cerberus/jwt/CerberusJwtClaims.java:
```diff
@@ -0,0 +1,121 @@
+import com.nike.cerberus.domain.AuthTokenInfo;
+import java.time.OffsetDateTime;
+
+public class CerberusJwtClaims implements AuthTokenInfo {
```
Could definitely use the Lombak getters and setters in this class

shawn-sher replied:
I'll add Getters

melanahammel commented on cerberus-web/src/main/java/com/nike/cerberus/jwt/CerberusJwtClaims.java:
```diff
@@ -0,0 +1,121 @@
+
+  @Override
+  public String getTokenHash() {
+    return null;
```
Why is this null for the CerberusJwtClaims?

melanahammel commented on cerberus-web/src/main/java/com/nike/cerberus/jwt/CerberusSigningKeyResolver.java:
```diff
@@ -0,0 +1,296 @@
+
+  private static final String DEFAULT_ALGORITHM = "HmacSHA512";
+  private static final String DEFAULT_JWT_ALG_HEADER = "HS512";
+  private static final int DEFAULT_MINIMUM_KEY_LENGTH = 512 / 8; // hardcoding these for now
```
When would we change these?

melanahammel commented on cerberus-web/src/main/java/com/nike/cerberus/jwt/JwtSecret.java:
```diff
@@ -0,0 +1,69 @@
+
+package com.nike.cerberus.jwt;
+
+public class JwtSecret {
```
Again: getters and setters

melanahammel commented on cerberus-web/src/main/java/com/nike/cerberus/jwt/JwtSecretData.java:
```diff
@@ -0,0 +1,32 @@
+
+/** A POJO that represents the JWT config */
+public class JwtSecretData {
+  private LinkedList<JwtSecret> jwtSecrets = new LinkedList<>();
```
I believe you can just add @Getter and @Setter to this line and delete the rest of the file

melanahammel commented on cerberus-web/src/main/java/com/nike/cerberus/service/AuthTokenService.java:
```diff
@@ -75,45 +99,112 @@ public CerberusAuthToken generateToken(
+    }
 
     OffsetDateTime now = OffsetDateTime.now();
+    // TODO: break up this if for two different messages (if present vs expired)
```
Is this TODO still necessary?

melanahammel commented on cerberus-web/src/main/java/com/nike/cerberus/service/JwtService.java:
```diff
@@ -0,0 +1,192 @@
+   * @return Does the token look like a JWT
+   */
+  public boolean isJwt(String token) {
+    return StringUtils.countMatches(token, SEPARATOR_CHAR) == 2;
```
What does 2 represent in this case?

shawn-sher replied:
There are three parts separated by '.' characters if token is a JWT token.

shawn-sher replied:
e.g. eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJKb2UifQ.ipevRNuRP6HflG8cFKnmUPtypruRC4fb1DWtoLL62SY from https://www.jsonwebtoken.io/

shawn-sher replied:
I'll make a constant for it.

melanahammel commented on cerberus-web/src/test/java/com/nike/cerberus/service/AuthTokenServiceTest.java:
```diff
@@ -93,6 +110,8 @@ public void before() {
     assertEquals(
         "The newly created token should have a refresh count of 0", 0, token.getRefreshCount());
 
+    // TODO: does this need to be fixed?
```
Is this TODO still relevant?

shawn-sher replied:
It's gone

melanahammel commented on cerberus-web/src/test/java/com/nike/cerberus/service/AuthTokenServiceTest.java:
```diff
@@ -105,9 +124,55 @@ public boolean matches(Object argument) {
+    assertEquals(
+        "The newly created token should have a refresh count of 0", 0, token.getRefreshCount());
+
+    // TODO: does this need to be fixed?
```
Is this TODO still relevant?

shawn-sher replied:
It's gone

melanahammel commented on cerberus-web/src/test/java/com/nike/cerberus/service/AuthTokenServiceTest.java:
```diff
@@ -167,16 +257,69 @@ public void test_that_when_a_token_is_expired_empty_is_returned() {
+    String id = UUID.randomUUID().toString();
+    String tokenId = "abc.123.def";
+    OffsetDateTime now = OffsetDateTime.now();
+    //    final String fakeHash = "kjadlkfjasdlkf;jlkj1243asdfasdf";
```
This comment can be deleted

shawn-sher replied:
It's gone

melanahammel commented on cerberus-web/src/test/java/com/nike/cerberus/service/AuthTokenServiceTest.java:
```diff
@@ -167,16 +257,69 @@ public void test_that_when_a_token_is_expired_empty_is_returned() {
+    assertEquals(0, token.getRefreshCount());
+  }
+
+  //   TODO: does this need to be updated to use jwt?
```
Another TODO?

Commit:
wip jwt fixes

Commit:
fix spinnaker dep path

Commit:
make s3 bucket property optional

Commit:
fix test

shawn-sher committed:
Switched to scrollbox for viewing token

Commit:
testing built-in jwt compression

shawn-sher committed:
Added check for token length

Added exception if the token is too large
Updated logging to include principal

shawn-sher committed:
Renamed constant

shawn-sher committed:
Auth service can use session if jwt token too long

If the accept type is set to ALL and the JWT token would be too long,
then issue a session token instead of failing.

Commit:
fix jwt service tests

Commit:
fix updated unittest

Commit:
bump version

Commit:
update dependencies to fix failing build

slichlyter12 committed:
suppressed CVE warning for postcss

mayitbeegh commented on cerberus-web/build.gradle:
```diff
@@ -64,9 +64,17 @@ dependencies {
 
+
+  // JWT
+  implementation "io.jsonwebtoken:jjwt-api:0.10.5"
```
These versions are quite old and they fixed a few vulnerabilities since then. Let's at least upgrade to 0.10.8 or newer.

mayitbeegh commented on cerberus-web/src/main/java/com/nike/cerberus/service/AuthTokenService.java:
```diff
@@ -75,45 +99,124 @@ public CerberusAuthToken generateToken(
+        } catch (ApiException e) {
+          String tooLongName = DefaultApiError.AUTH_TOKEN_TOO_LONG.getName();
+          boolean isTooLong =
+              e.getApiErrors().stream().anyMatch(err -> err.getName().equals(tooLongName));
```
This is a bit fragile and not type safe. If these lines are going to be removed later then that's fine. Otherwise, consider changing JwtService.java:110 into throwing a checked exception.

shawn-sher committed:
Updated io.jsonwebtoken dependencies

mayitbeegh commented on cerberus-web/src/main/resources/cerberus.yaml:
```diff
@@ -232,6 +232,21 @@ cerberus:
+      # Every minute
+      cronExpression: "0 * * ? * *"
+
+    jwtBlacklistCleanUpJob:
```
Change this to jwtBlocklistCleanUpJob

mayitbeegh replied:
Also I think this should be disabled by default

shawn-sher replied:
Won't the database grow without bound though?

mayitbeegh commented on cerberus-web/src/main/resources/cerberus.yaml:
```diff
@@ -232,6 +232,21 @@ cerberus:
+      # Every 5 seconds
+      cronExpression: "0/5 * * ? * *"
+
+    jwtBlacklistRefreshJob:
```
Change this to jwtBlocklistRefreshJob

mayitbeegh replied:
Also should be off by default

mayitbeegh commented on cerberus-web/src/main/java/com/nike/cerberus/record/JwtBlocklistRecord.java:
```diff
@@ -0,0 +1,46 @@
+import java.time.OffsetDateTime;
+import lombok.Data;
+
+@Data
```
This class needs some cleanup. Either delete the setters or delete the lombok annotation

shawn-sher committed:
Updated to address PR comments

shawn-sher committed:
Updated to address PR comments

Trimmed down AuthTokenInfo interface
Cleaned up comments
Set defaults for jwt jobs
Added jwt related settings to cerberus.yaml
Made clarifications to some constants
Updated logging levels
Added logging to get rid of a TODO message

Commit:
add tests for AuthTokenTooLongException
